### ISSUE TITLE
Object is possibly 'undefined'

### ISSUE THREAD

**Issue by troy351** [microsoft/TypeScript#29642]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:** Object undefined

**Code** turn `strictNullChecks` on

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
type EventType = 'click' | 'dblclick'

const handlerMap: { [P in EventType]?: any[] } = {}

function addHandler<P extends EventType>(evType: P) {
  const handlerList = handlerMap[evType] || []
  handlerList.push({}) // Error here: Object is possibly 'undefined'
  handlerMap[evType] = handlerList
}
```

**Expected behavior:** No Error

**Actual behavior:** Shows Error

**Playground Link:**  http://www.typescriptlang.org/play/#src=type%20EventType%20%3D%20'click'%20%7C%20'dblclick'%0Aconst%20handlerMap%3A%20%7B%20%5BP%20in%20EventType%5D%3F%3A%20any%5B%5D%20%7D%20%3D%20%7B%7D%0Afunction%20addHandler%3CP%20extends%20EventType%3E(evType%3A%20P)%20%7B%0A%20%20const%20handlerList%20%3D%20handlerMap%5BevType%5D%20%7C%7C%20%5B%5D%0A%20%20handlerList.push(%7B%7D)%20%2F%2F%20Error%20here%3A%20Object%20is%20possibly%20'undefined'%0A%7D%0A

**Related Issues:**

**Comment by collin5** [comment]
Will be giving this a try.

**Comment by ktoto** [comment]
It's the same issue?
```typescript
interface State {
    [i: string]: {data?: {id:string}} | undefined
 }

declare var x: State;

const id = 'xx';
const y = x[id] && x[id].data; // Error: Object is possibly 'undefined' (with strictNullChecks
const yId = y && y.id; // correct type – string | undefined
```

Playground link : https://www.typescriptlang.org/play/#src=interface%20State%20%7B%0D%0A%20%20%20%20%5Bi%3A%20string%5D%3A%20%7Bdata%3F%3A%20%7Bid%3Astring%7D%7D%20%7C%20undefined%0D%0A%20%7D%0D%0A%0D%0Adeclare%20var%20x%3A%20State%3B%0D%0A%0D%0Aconst%20id%20%3D%20'xx'%3B%0D%0A%0D%0Aconst%20y%20%3D%20x%5Bid%5D%20%26%26%20x%5Bid%5D.data%3B%0D%0A%0D%0Aconst%20yId%20%3D%20y%20%26%26%20y.id%3B%0D%0A

**Comment by weswigham** [comment]
#29317 should be the fix here (which I have mentioned in @collin5 's PR). Specifically, the type of `a || b` should be calculated as `(typeof a & not FalseyTypes) | typeof b`, where `FalseyTypes` is `0 | false | "" | undefined | null`.

@RyanCavanaugh given that, should we move this out of `3.4`, since we won't be shipping them in 3.4?

**Comment by weswigham** [comment]
@ktoto your issue is different - since `State` has an index signature, we don't track refinements on individual property lookups within it - see #17960 and #29042 (and go updoot those issues so they don't sit in the backlog).

**Comment by MufidJamaluddin** [comment]
I have same issue!

![undefined](https://user-images.githubusercontent.com/28973632/57500557-01e48280-730e-11e9-9e73-5638effb2a6b.JPG)

**Comment by MufidJamaluddin** [comment]
I have same issue. :(

**Typescript Version** : 3.4.5

![undefined](https://user-images.githubusercontent.com/28973632/57501071-41ac6980-7310-11e9-87b5-2954543862ab.JPG)

**Comment by troy351** [comment]
@MufidJamaluddin  duplicate of #7719

**Comment by micky2be** [comment]
Same issue @MufidJamaluddin 
I have a checker in place to avoid duplicating code but Typescript ignores it and output this error.

**Comment by marcj** [comment]
![Screenshot 2019-08-06 at 15 45 00](https://user-images.githubusercontent.com/450980/62545020-298dad00-b861-11e9-936e-e0f70b317df7.png)

I got the exact same issue since upgrading to newer version. I honestly don't know what to do as this worked for ages.

**Comment by MBerka** [comment]
@marcj Same error, but as with @ktoto's, the issue has more to do with https://github.com/microsoft/TypeScript/issues/29042 and https://github.com/microsoft/TypeScript/issues/17960. Ultimately, the question is how much TypeScript can be asked to simulate when checking keys. 3.5.3 seems to handle checking of a constant string index whose value was set with a constant string index after initialization, and of a constant variable index whose value was defined on declaration, but not a variable index for a value assigned after initialization.
```typescript
const myObj: { prop?: number } = {};
myObj['prop'] = 1;
if (myObj['prop']) {
    myObj['prop'] = 1 * myObj['prop']; // No error
}
const field = 'prop'
if (myObj[field]) {
    myObj[field] = 1 * myObj[field]; // Possibly undefined
}
const otherObj = {prop: 1};
if (otherObj[field]) {
    otherObj[field] = 1 * otherObj[field]; // No error
}
```
Your case stands out because it requires dynamic assignment, dynamic retrieval, and continued access to the original object. Casting workaround that could get tedious:
```typescript
interface KeyType {setValue: (input: string) => void};
const myObj: { aa?: KeyType} = {};
const key = 'aa';
myObj[key] = {setValue: (input): void => {console.log(input);}}

if (myObj[key]) {
    (myObj[key] as KeyType).setValue('bb');
}
```

**Comment by npenin** [comment]
same issue for me with this code: https://www.typescriptlang.org/play/index.html?target=6#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgYQgFsjMkATAKAG9K565sIl0EBzALgOdbYFcomGAmYBnANx0GyAFbBsMAPxdRMKMjYBtALqSAvpUqhIsRCjRZccABIwYYQi3YChIpHFCoKo7k-6DhZhopeiJgGAALCHIVNQ1JBjgoCD5UWPUkNn1DY2h4ZFQMHDxHXhdApFFgxM0Aa2AAT3SNbS4+CmBWJGByOAAfXzKAtwSGCLswZXpbe1LnYeZ9SSNwPLNCyxKeedcg2kTZeSVmzJ1swwA3TCg4TA5CEjJyAF5qRm3Oaj09SSubsGSYDQMAa1ggABtyGhXgCIEDYA0AHKYMIcADkqFUaJ+hgQ6AAFJg4AAyYm3AB0TD8mlh8JBYMhaHJtOBSJRwG0JLJmEpHxpgNZDKhUGZAoRyLC2nJhwUAEpgnogA

```
export interface Command
{
    config: Configurations;
    inject?: string[];
}

export interface HttpConfiguration extends Configuration
{
    method: string;
    route: string;
}

export interface Configurations
{
    [key: string]: undefined | Configuration;
    http?:  HttpConfiguration;
};

export interface Configuration
{
    inject?: string[];
}


var a:Command={ config:{}};
var propertyHolder={propertyName:'test'};

if(a && a.config[propertyHolder.propertyName] && a.config[propertyHolder.propertyName].inject)
{
}
```

**Comment by chamie** [comment]
Another example without dynamic property access:
```ts
import React from "react";

interface IPropsPrivate<T> {
  someField: T;
  maybeRows?: string[];
}

type IProps<T> = T extends string ? never : IPropsPrivate<T>;

class Class1<T> extends React.PureComponent<IProps<T>> {
  public render() {
    const maybeRows = this.props.maybeRows;
    const notUndefined = maybeRows === undefined ? [] : maybeRows;

    notUndefined.filter(() => true); // Error: Object is possibly 'undefined'.

    const twiceGuaranteedUndefined = notUndefined || [];

    twiceGuaranteedUndefined.filter(() => true); // Error: Object is possibly 'undefined'.

    return null;
  }
}

// This works fine though
class Class2<T> extends React.PureComponent<IProps<T>> {
  public render() {
    const { maybeRows = [] } = this.props;

    maybeRows.filter(() => true);

    return null;
  }
}
```
[>>Playground](http://www.typescriptlang.org/play/index.html?esModuleInterop=true&target=2&jsx=2#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4Aoc4AOxiSk3STgEkAFHMAZw+ADcUdADwAVAHxwA3uThwuuJADFgSADYATAFxwRFWSBQBPAEZIEEAO5cA-Nq4woNAOYBtALoUAvpRiGwzdk4uUQkAXh04JAAPOmp1LjkHZzhrOGokPno4bUCIbl4BYXEKcjRVFC4EgGFyyoBGEMiYpDiE5HQYADo2AFciKtxIdNohXO4QiWlZMB7jVWA0OCI4+gAKAEopGVk4NAhqezgDEzNLBPCYAAtgLk6wIM7j03MrPR29g-hqCBgAVRXMDQkOo4OEnqcrKDQuEegCgSDUu5skcjM8ziUdmkfv91EhAel1J1Aao6FBVhtQRIHD0kOtSHAAPQMuAAUSgOCg2gA8sYAFZIDBwG5wSCVYBzQxwADksNx+OBUs6lExH0OMAsCyQAHEeigoChaEhgTi8fDQVi-nCCXAAD42uDuDE7dWanV6g10Y1W4FE4AktYU0JUqA0umM5lsjncvkC+DC0VccWqSUy73qRXKnZEGB9ahpHqqVRvbzechMnTXBIWaAAawS8rgVwgPScl1KtWqHYATI1orF4ohiF1ev1BvsWjBRhw8sFxJNtjM5gsli1cWTNlN3vtDpIUScXucHW44J5zVcbncgk79KiIbdiaTyZsg42Q7Tryuc1A89QC0XtiW5BAA)

**Comment by joelalejandro** [comment]
Another example, tried on `3.7.0-dev.20190928`:

```ts
type DialogProps = {
  buttons?: {
    primary: boolean;
    secondary?: boolean;
  };
};

function Dialog({ buttons }: DialogProps) {
    const atLeastOneButton = buttons !== undefined;

    if (atLeastOneButton) {
      const label = buttons.primary; 
      // throws "Object is possibly 'undefined' over `buttons`
    }
}
```

Apparently, the compiler doesn't understand that `atLeastOneButton` is performing a not-undefined check for the same `buttons` variable.

**Playground link**
https://www.typescriptlang.org/play/?ts=Nightly#code/C4TwDgpgBAIglgQwDYHsDmAFATisBnKAXigG8AoKKAIwFdhgUA7PAfgC5SLKows4BbBFhAcqKFEggJGAbi6U8EAMZMAJkJDtq4ydLmUAvnKNkyAMxqMlwOE1iJUaABQlqdBsygGO8ZOmy4eACUnNxQKszAUAjAADJSeMAA8owQAELudsS09EwEAISExJaqEGZwqapQAGTVbrnMAHS8AhpQhcWMpeWVcvJQcGZQTjHxCIkp6ZmMIeRhlBGJUEgIVBBIRPUeeM18gsL63AZkBkA

**Comment by younesmln** [comment]
another example

```
class Test {
    num?: number;

    constructor(num?: number) {
        this.num = num
    }

    checkIsPositive() {
        return this.num && this.num > 0
    }

    doStuff() {
        if (!this.checkIsPositive()) return;
        console.log(this.num ** 4);
    }
}
```

playground link: http://www.typescriptlang.org/play/?ssl=16&ssc=2&pln=1&pc=1#code/MYGwhgzhAEAqCmEAu0DeAoaXoDsCuAtgPwBcuhARvAE4Dc6m2wA9jstXsEs9QBT7EyAqtQCUaRtmxIAFgEsIAOgHQAvOQKToAXwZTgM+MADWASQgAFZhDlI5AN3i9xGKVOrwkeajmiyFyoTQAGTBfvJKKgB80AAMWrpaACbMAMpeAGYZzhJu2HIZ0LwAhP5KBkZmlta2Dk6i4h5ePvR5WCxszCDwiiDMAOa8ZYEE0ABUY9AALKKt2LqJ6EA

**Comment by rasmus-storjohann-PG** [comment]
One more example: 
```ts
export function* updateTaskServices(action: actions.BuildServicesRequestAction): UpdateResult {
    const topicId = action.payload.topicId;
    try {
        const deviceLocationResponse = yield call(getDeviceLocation, action.payload.manualUserLocation);

        if (deviceLocationResponse && deviceLocationResponse.type) { // <---- error on this line
            if (errors.isNoLocationPermissionError(deviceLocationResponse)) {
                return yield put(actions.buildServicesErrorAction(topicId, deviceLocationResponse.type));
            }

            if (errors.isLocationFetchTimeoutError(deviceLocationResponse)) {
                return yield put(actions.buildServicesErrorAction(topicId, deviceLocationResponse.type));
            }
        }
        const apiResponse: APIResponse = yield call(searchServices, topicId, deviceLocationResponse);
        if (errors.isBadResponseError(apiResponse)) {
            return yield put(actions.buildServicesErrorAction(topicId, Errors.BadServerResponse));
        }

        const validatedApiResponse = validateServicesAtLocationArray(apiResponse.results);
        if (!validatedApiResponse.isValid) {
            return yield put(actions.buildServicesErrorAction(topicId, Errors.InvalidServerData));
        }
}
```
with this error message:
```
src/sagas/services.ts:21:39 - error TS2532: Object is possibly 'undefined'.

21         if (deviceLocationResponse && deviceLocationResponse.type) {
```

**Comment by rasmus-storjohann-PG** [comment]
One more example, this is pretty broken:
```ts
export function* updateTaskServices(action: actions.BuildServicesRequestAction): UpdateResult {
    const deviceLocationResponse = yield call(getDeviceLocation, action.payload.manualUserLocation);

    if (deviceLocationResponse && deviceLocationResponse.type) {
        if (errors.isLocationFetchTimeoutError(deviceLocationResponse)) {
            return yield put(actions.buildServicesErrorAction('', deviceLocationResponse.type));
        }
    }
    yield put(actions.buildServicesSuccessAction('', R.map(serviceFromValidatedJSON, deviceLocationResponse.validData)));
}
```
error message, deviceLocationResponse can definitely not be undefined right after `deviceLocationResponse && ...`:
```
src/sagas/services.ts:17:35 - error TS2532: Object is possibly 'undefined'.

17     if (deviceLocationResponse && deviceLocationResponse.type) {
```

**Comment by weswigham** [comment]
@rasmus-storjohann-PG gunna need some more information on the shapes of the types in question to get an actual repro of that - I'd also open a new issue, since it looks like a pretty distinct bug (just resulting in the same diagnostic message).

**Comment by itmayziii** [comment]
Another example below in case it helps.

http://www.typescriptlang.org/play/?noImplicitReturns=false#code/JYOwLgpgTgZghgYwgAgArAQawK4AcAickAzsgN4BQyyA2phAJ4BcyxYUoA5gLotm31myENgC2AI2i9khSMgC+yAD7JsIACYQYoCOoryKVZDDUIwwAPYhjodeix5ZKABS4MOAkQjEW9j0+IAGmQ4dXUobx9Wdi5g2wgADxYRCWgAShYnZVUNLR11ciNqYBhkZwBCNwdPEhpQ8MjuZRVK90cvYjqwiOJibhp4hO40wuox5AiwbCgQIoVDccnp6yr-Dq6G3v7B7iMDIA

**Comment by gkamperis** [comment]
why is this not allowed?

```
checkIsPositive(num: number | undefined): boolean {
        return num > 0
    }    

```
[playground](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=6&pc=1#code/MYGwhgzhAEAqCmEAu0DeAoaXrABb2AGsBJCABQHsIBLJagN3gAoA7AVwFsAuadjgI3gAnaAB9obFgBN4AM2ot4UgJQ9+FCiHhgWaTNgND4SNkN19oAPmgAGfVgC+B9A-RA)

**Comment by MBerka** [comment]
> why is this not allowed?
> 
> ```
> checkIsPositive(num: number | undefined): boolean {
>         return num > 0
>     }    
> ```
> 
> [playground](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=6&pc=1#code/MYGwhgzhAEAqCmEAu0DeAoaXrABb2AGsBJCABQHsIBLJagN3gAoA7AVwFsAuadjgI3gAnaAB9obFgBN4AM2ot4UgJQ9+FCiHhgWaTNgND4SNkN19oAPmgAGfVgC+B9A-RA)

Because it does not make sense (in a strongly typed language) to supply a non-numerical value to an operation meant to compare two numbers. JavaScript's answer (always false if either argument cannot be converted into a number) is arbitrary. This is correct behavior unrelated to this issue.

**Comment by gkamperis** [comment]
@MBerka I disagree. I specifically typed the argument. I know it can be undefined. The comparison is valid JS why am I prevented from running?

**Comment by troy351** [comment]
@gkamperis That's valid JS but bad part of JS which TS whats to get rid of.
Since `undefined > 0 === false` doesn't mean `undefined` is smaller than `0`, right?

**Comment by gkamperis** [comment]
@troy351 `num ? (num > 0) : false` does not mean num is smaller than 0 either. I do not see the relevance.

This check is about NPEs only. There is no chance an NPE can happen here. So the message is not valid.

**Comment by troy351** [comment]
@gkamperis Yeah, for the complier, `num !== undefined ? (num > 0) : false` and `num > 0` comes out the same result. But for progammers, it's obvious the first one is better.

TS was made for JS users to make less mistakes and that's how TS did it.

Consider this one
```javascript
checkIsPositive(num: number | undefined): boolean {
        return !(num <= 0)
}   
```

**Comment by RyanCavanaugh** [comment]
Tracking at #7719

**Comment by andreialecu** [comment]
@RyanCavanaugh did you link the wrong issue? That one has been closed since 2018.

**Comment by impulse** [comment]
Also having this issue

**Code**
```ts
type Key = "key1" | "key2";
type Keys = {
  [P in Key]?: { keyProp: string };
};

const keys: Keys = {
  key1: { keyProp: "keyPropVal" },
};

let someKey!: Key;

if (keys[someKey]) {
  // This shows Object is possibly 'undefined'
  keys[someKey].keyProp;
}
```



<details><summary><b>Output</b></summary>

```ts
"use strict";
const keys = {
    key1: { keyProp: "keyPropVal" },
};
let someKey;
if (keys[someKey]) {
    // This shows Object is possibly 'undefined'
    keys[someKey].keyProp;
}

```


</details>


<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```


</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/index.html?ssl=15&ssc=2&pln=1&pc=1#)

**Comment by alokadhao20** [comment]
Hello All,

> The object is possibly 'undefined',

This warning is shown by the editor when we provide the body definition of some variable in the form of the interface and add some of the fields as optional 

Eg:- 
interface Blog {
  autherName: string;
  contents?: any[],
  id: string
}

While using this interface Blog in our code if we are trying to do some loop on Blog.contents 
Eg:- 
const newBlog: Blog = JSON.parse(blog);
for (const content of newBlog.contents) {
   console.log("content ", content );
}

This will error in for loop declaration for (const content of newBlog.contents) because we have said in an interface that contents property in Blob object can be or cannot be there hence we first need to check in code if the property is present then we should use it. As follows 

if (newBlog.contents && newBlog.contents.length > 0) {
    for (const content of newBlog.contents) {
      console.log("content ", content );
   }
}

This is one way to resolve this warning.

Happy Coding!

**Comment by harishkshetty** [comment]
> Hello All,
> 
> > The object is possibly 'undefined',
> 
> This warning is shown by the editor when we provide the body definition of some variable in the form of the interface and add some of the fields as optional
> 
> Eg:-
> 
> ```
> interface Blog {
>   autherName: string;
>   contents?: any[],
>   id: string
> }
> ```
> 
> While using this interface Blog in our code if we are trying to do some loop on Blog.contents
> Eg:-
> 
> ```
> const newBlog: Blog = JSON.parse(blog);
> for (const content of newBlog.contents) {`
>    console.log("content ", content );
> }
> ```
> 
> This will error in for loop declaration for (const content of newBlog.contents) because we have said in an interface that contents property in Blob object can be or cannot be there hence we first need to check in code if the property is present then we should use it. As follows
> 
> Eg:-
> 
> ```
> if (newBlog.contents && newBlog.contents.length > 0) {
>     for (const content of newBlog.contents) {
>       console.log("content ", content );
>    }
> }
> ```
> 
> This is one way to resolve this warning.
> 
> Happy Coding!

Thanks alokadhao20,It solved issue for me.

**Comment by subins2000** [comment]
@RyanCavanaugh did you link the wrong issue ? That one has been closed since 2018.

**Comment by RyanCavanaugh** [comment]
Locking because everyone is posting about all of their "is possibly undefined" code snippets here without context for the original post.

The inability to handle `x || []` when `x` is a type parameter is a current design limitation wherein we need to have some new mechanism for describing higher-order facts about generic type parameters; today if something is `T` and we know it's truthy or falsy, there's no mechanism for us to better describe its assignability to any given target type. Everything we've tried so far here has been a performance or usability disaster so it's on the back burner for the time being until we can hopefully have some better insight on how to represent this in a more ergonomic/efficient way.
